cleverhans/attacks.py

```diff
@@ -1,62 +1,233 @@
-    :param x: the input
-    :param predictions: the model's output
-    :param eps: the epsilon (input variation parameter)
-    :param back: switch between TensorFlow ('tf') and
-                Theano ('th') implementation
-    :param clip_min: optional parameter that can be used to set a minimum
-                    value for components of the example returned
-    :param clip_max: optional parameter that can be used to set a maximum
-                    value for components of the example returned
-    :return: a tensor for the adversarial example
-    """
-    if back == 'tf':
-        # Compute FGSM using TensorFlow
-        from .attacks_tf import fgsm as fgsm_tf
-        return fgsm_tf(x, predictions, eps, clip_min=clip_min,
-                       clip_max=clip_max)
-    elif back == 'th':
-        # Compute FGSM using Theano
-        from .attacks_th import fgsm as fgsm_th
-        return fgsm_th(x, predictions, eps, clip_min=clip_min,
-                       clip_max=clip_max)
-
-
-def jsma(sess, x, predictions, grads, sample, target, theta, gamma=np.inf,
-         increase=True, back='tf', clip_min=None, clip_max=None):
-    """
-    A wrapper for the Jacobian-based saliency map approach.
-    It calls the right function, depending on the
-    user's backend.
-    :param sess: TF session
-    :param x: the input
-    :param predictions: the model's symbolic output (linear output,
-        pre-softmax)
-    :param sample: (1 x 1 x img_rows x img_cols) numpy array with sample input
-    :param target: target class for input sample
-    :param theta: delta for each feature adjustment
-    :param gamma: a float between 0 - 1 indicating the maximum distortion
-        percentage
-    :param increase: boolean; true if we are increasing pixels, false otherwise
-    :param back: switch between TensorFlow ('tf') and
-                Theano ('th') implementation
-    :param clip_min: optional parameter that can be used to set a minimum
-                    value for components of the example returned
-    :param clip_max: optional parameter that can be used to set a maximum
-                    value for components of the example returned
-    :return: an adversarial sample
-    """
-    if back == 'tf':
-        # Compute Jacobian-based saliency map attack using TensorFlow
-        from .attacks_tf import jsma_tf
-        return jsma_tf(sess, x, predictions, grads, sample, target, theta,
-                       gamma, increase, clip_min, clip_max)
-    elif back == 'th':
-        raise NotImplementedError("Theano jsma not implemented.")
+class Attack:
+    """
+    Abstract base class for all attack classes.
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, x, pred, y=None, backend='tf', clip_min=None, clip_max=None):
+        """
+        :param x: A placeholder for the model inputs.
+        :param pred: The model's symbolic output.
+        :param y: A placeholder for the model inputs. Only provide this parameter if
```

Comment:
This param doesn't look like it's necessary for all attacks

Thoughts @goodfeli?